[STEP 4] 로또(수동)

y2gcoder opened the pull request:
안녕하세요 윤혁님! 

또 윤혁님의 피드백을 받고자 PR을 올리게 되었습니다 :) 

### 작업 내역
- STEP 3 때 말씀해주셨던 피드백 반영
- STEP 4 기능 요구사항 반영
- STEP 4 프로그래밍 요구사항 반영

### 질문
- 상수 사용: 티켓 가격을 공통적으로 사용하는 객체가 2곳 있어 고민을 하다 TICKET_PRICE 라는 이름으로 LottoConstants 라는 object를 만들어 상수로 빼게 되었습니다. 윤혁님은 여러 객체에서 사용하는 공통된 값이 있으면 어떻게 처리하시는지 궁금합니다!

윤혁님 덕분에 코틀린에 대한 이해도 조금씩 늘어가고, 꼼꼼하게 객체지향적으로 짜게 되는 것 같습니다!

항상 피드백 감사히 받고 있습니다! 



양영근 올림.

y2gcoder committed:
feature(lotto): STEP 3 리뷰어님(윤혁님) 피드백 반영

LottoNumber: 유효성 검사 메시지에 입력값 포함
LottoTicket: matchesBonusBall()을 contain() 에 통합
LottoTicket: 부생성자 생성(vararg numbers: Int)
LottoTicket: numbers를 내부에서 Set으로 관리
LotteryWinningChecker: checkTicketAll() 구현
LotteryWinningChecker: 좀 더 적합한 이름으로 변경 ex) WinningTicket

y2gcoder committed:
feature(lotto): 금액에 대한 value class Money 추가

확장함수와 operator fun 을 이용해서 최대한 Int를 사용할 때와 비슷하게 사용할 수 있도록 구현

y2gcoder committed:
fix(lotto): 2등 당첨 통계 출력 수정

- ', 보너스 볼 일치' 문구 추가
- Money 교체 후 출력 텍스트 변동 없도록 반영

y2gcoder committed:
feature(lotto): 로또 티켓 가격을 상수로 변환

LottoTicketMachine 과 WinningBoard 에서 사용 중이기 때문에 상수화해서 재사용할 수 있도록 관리

y2gcoder committed:
feature(lotto): 수익률을 Returns로 래핑

수익률 또한 value class(Returns)로 래핑 및 java.math 패키지 사용하지 않고 코틀린 표준 라이브러리 사용해서 소수점 자리 + 버림 표현

y2gcoder committed:
feature(lotto): 로또 티켓 머신에 수동 티켓 발급 기능 추가

LottoTicketMachine은 수동으로 lottoNumbers 를 받아 로또 티켓을 발급할 수 있다

y2gcoder committed:
feature(lotto): 로또 티켓 머신에 발급 가능한 티켓 수 구하는 함수 추가

- LottoTicketMachine.affordableTickets(): Int
- 계산 : (잔액 / 티켓 1매 가격)

y2gcoder committed:
feature(lotto): 로또 프로그램에 수동 로또 생성 UI 추가

malibinYun commented on src/main/kotlin/lotto/ConsoleInput.kt:
```diff
@@ -2,14 +2,30 @@ package lotto
         }
 
-        fun inputWinningNumbers(): List<LottoNumber> {
+        fun inputManualLottoNumbers(
+            inputManualTicketCount: Int,
+            lotteryTicketMachine: LotteryTicketMachine,
+        ): List<LottoTicket> {
+            println("수동으로 구매할 번호를 입력해 주세요.")
```
ConsoleInput을 받는 부분에서 LotteryTicketMachine 객체 의존이 필요한 것에 조금 어색하다고 느껴요.
입력값 객체는 입력 값만 받아와서 반환하기만 하고, 그 후 가공은 main함수나 혹은 다른 객체에서 수행하도록 만드는 건 어떨까요?

malibinYun commented on src/main/kotlin/lotto/LottoConsole.kt:
```diff
@@ -4,20 +4,30 @@ fun main() {
         )
     println()
 
-    val checkedResults = lotteryTickets.map { lotteryWinningChecker.checkTicket(it) }
+    val checkedResults = lotteryTickets.map { winningTicket.checkTicket(it) }
```
WinningTicket에 만들어 둔 checkTicketAll 메서드를 활용해볼 수 있겠어요!

malibinYun commented on src/main/kotlin/lotto/LottoConstants.kt:
```diff
@@ -0,0 +1,5 @@
+package lotto
+
+object LottoConstants {
+    val TICKET_PRICE = Money(1000)
+}
```
> 상수 사용: 티켓 가격을 공통적으로 사용하는 객체가 2곳 있어 고민을 하다 TICKET_PRICE 라는 이름으로 LottoConstants 라는 object를 만들어 상수로 빼게 되었습니다. 윤혁님은 여러 객체에서 사용하는 공통된 값이 있으면 어떻게 처리하시는지 궁금합니다!

저는 다른 개발자들이 해당 상수를 찾을 때, 어떤 객체를 가장 먼저 찾아볼까? 를 먼저 고민해 봐요.
상수를 상수 object에 모두 몰아 넣는 것이 컨벤션이라면 따라야겠지만, 그렇지 않은 경우에 객체에 위치 시키는 것을 선호해요.

로또 티켓 하나의 가격이니, 로또 티켓에 있을까?
로또의 가격은 발행하는 (혹은 판매하는) 곳의 역할이니, 이곳에 가격이 있지 않을까?
이런 생각에 가장 연관 깊어 보이는 곳과 제 주관이 조금 섞여서 판단한 위치에 상수를 두는 편이에요.

완전 부자연스러운 곳에 있지 않는 이상 티켓에 상수가 있든, 발매기에 상수가 있든 결국은 찾을 것이라 생각해요.
완전 부자연스러운 곳이라면... WinningBoard에 상수가 존재하는 경우 정도가 떠오르네요! 이런 경우에는 상수를 찾기 매우 어려울 것으로 예상해요.

malibinYun commented on src/main/kotlin/lotto/LottoTicket.kt:
```diff
@@ -1,25 +1,28 @@
 
 class LottoTicket(numbers: List<LottoNumber>) {
-    val numbers: List<LottoNumber> = numbers.sorted()
+    private val _numbers: Set<LottoNumber> = numbers.toSet()
+
```
[.toSortedSet()](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/to-sorted-set.html) 을 활용해봐도 좋겠네요!

malibinYun commented on src/main/kotlin/lotto/Money.kt:
```diff
@@ -0,0 +1,28 @@
+        return amount.toDouble()
+    }
+
+    companion object {
+        val ZERO = Money(0)
+    }
```
자주 사용하는 값을 상수화 해주셨네요! 👍

malibinYun commented on src/main/kotlin/lotto/Returns.kt:
```diff
@@ -0,0 +1,24 @@
+import lotto.LottoConstants.TICKET_PRICE
+import kotlin.math.floor
+import kotlin.math.pow
+
+class Returns(totalWinningAmount: Money, totalCost: Money) {
```
Returns 라는 이름을 보고 어떤 일을 하는 객체인지 유추하기 어려웠어요.
어떤 일을 하는지 조금 더 자세히 이름에 녹여내보는 건 어떨까요?

y2gcoder replied:
RateOfReturn 이라는 이름으로 바꿔봤습니다! 수익률이 Returns 혹은 RateOfReturn이라는 이름으로 많이 사용되는 것 같아서 그렇게 해봤습니다!

malibinYun commented on src/main/kotlin/lotto/Money.kt:
```diff
@@ -0,0 +1,28 @@
+package lotto
+
+@JvmInline
+value class Money(val amount: Int) : Comparable<Money> {
```
1등이 두 번 당첨되는 경우 오버플로가 발생할 수 있으니, Long으로 설정해보는 건 어떨까요?

malibinYun reviewed (changes requested):
4단계 미션 고생 많으셨어요!
요구사항을 너무 잘 반영해주셔서 크게 피드백 드릴 부분이 없었네요 👍
더 반영하고 싶으신 내용이 있으시다면 반영 후 리뷰 요청 부탁드리고,
이만 미션을 마무리 짓고 싶으시다면 코멘트 남겨주세요! 🙂

y2gcoder committed:
feature(lotto): 윤혁님 피드백 반영

- ConsoleInput.inputManualLottoNumbers(): lotteryTicketMachine에 대한 의존을 피하기 위해 List<LottoNumber> 를 반환하도록 변경 후 LottoConsole 에서 조립
- LottoConsole: winningTicket.checkTicketAll 을 사용하도록 변경
- LottoConstants: 삭제. TICKET_PRICE 는 LottoTicketMachine 내부의 companion object 에서 관리
- Return -> RateOfReturn 으로 이름 변경
- Money: 내부 값 타입 Long으로 변경(, 연산 관련된 부분도 Long으로 변경)

추가
- ResultView: 연산 기능 없이 화면에 표시해주는 로직만 남기도록 함
- LottoConsole: 수동 티켓 개수를 0개 이하로 입력하면 수동 티켓 번호 입력 건너뛰도록 처리

y2gcoder commented:
윤혁님의 피드백을 반영해서 다시 리뷰 요청 드리고 싶습니다! 
감사합니다 :)

malibinYun commented on src/main/kotlin/lotto/LottoTicket.kt:
```diff
@@ -1,27 +1,34 @@
     init {
-        require(numbers.size == 6) {
+        require(numbers.size == LOTTO_NUMBERS_SIZE) {
             "로또 번호는 총 6개여야 합니다"
         }
-        require(numbers.size == numbers.distinct().size) {
+        require(numbers.size == _numbers.size) {
             "로또 번호는 모두 서로 다른 번호여야 합니다"
         }
```
에러 메시지에 입력 값을 담아 준다면 디버깅하기 더 편리할 것으로 기대돼요 :)

malibinYun reviewed (approved):
리뷰가 늦어져서 죄송합니다 😥
반영하신 피드백과 다시 한 번 프로젝트를 보았을 때 크게 피드백 드릴 부분이 없었네요!
로또 미션 모두 완수하시느라 정말 고생 많으셨어요 :)
남은 다른 미션들도 모두 완주하시길 응원할게요! 💪
